src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -2331,6 +2451,114 @@ private void Dispose(bool disposing)
+
+                foreach (var inputCacheFile in inputCacheFiles)
+                {
+                    var (configCache, resultsCache, exception) = DeserializeCaches(inputCacheFile);
+
+                    if (exception != null)
+                    {
+                        LogErrorAndShutdown(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("ErrorReadingCacheFile", inputCacheFile, exception.Message));
+                        return false;
+                    }
+
+                    cacheAggregator.Add(configCache, resultsCache);
+                }
+
+                var cacheAggregation = cacheAggregator.Aggregate();
+
+                // using caches with override (override queried first before current cache) based on the assumption that during single project cached builds
+                // there's many old results, but just one single actively building project.
+
+                _componentFactories.ReplaceFactory(BuildComponentType.ConfigCache, new ConfigCacheWithOverride(cacheAggregation.ConfigCache));
+                _componentFactories.ReplaceFactory(BuildComponentType.ResultsCache, new ResultsCacheWithOverride(cacheAggregation.ResultsCache));
+
+                return true;
+            }
+            catch
+            {
+                CancelAndMarkAsFailure();
+                throw;
+            }
+
+            (IConfigCache ConfigCache, IResultsCache ResultsCache, Exception exception) DeserializeCaches(string inputCacheFile)
+            {
+                try
+                {
+                    ConfigCache configCache = null;
+                    ResultsCache resultsCache = null;
+
+                    using (var fileStream = File.OpenRead(inputCacheFile))
+                    {
+                        var translator = BinaryTranslator.GetReadTranslator(fileStream, null);
+
+                        translator.Translate(ref configCache);
+                        translator.Translate(ref resultsCache);
+                    }
+
+                    ErrorUtilities.VerifyThrowInternalNull(configCache, nameof(configCache));
+                    ErrorUtilities.VerifyThrowInternalNull(resultsCache, nameof(resultsCache));
+
+                    return (configCache, resultsCache, null);
+                }
+                catch (Exception e)
+                {
+                    return (null, null, e);
+                }
+            }
+        }
+
+        private void LogErrorAndShutdown(string message)
+        {
+            var loggingService = ((IBuildComponentHost)this).LoggingService;
+
+            loggingService?.LogErrorFromText(
+                BuildEventContext.Invalid,
+                null,
+                null,
+                null,
+                BuildEventFileInfo.Empty,
+                message);
+
+            CancelAndMarkAsFailure();
+
+            if (loggingService == null)
+            {
+                // todo should we write this to temp file instead (like failing nodes do)
```

Comment:
Decided to throw because that's what other parts of the BeginBuild code do if logging throws: https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/BuildManager/BuildManager.cs#L424-L440

I could change all BeginBuild code to also dump the exception, but I would say changing this is beyond this PR. Looking at [DumpExceptionToFile](https://github.com/Microsoft/msbuild/blob/65845d8aaf94543978c2d57209add61d7b7d7944/src/Shared/ExceptionHandling.cs#L317), looks like there's a high risk the same exception gets appended multiple times as the stack unwinds. I'd rather fix that before dumping more exceptions from Begin/End Build, which will probably end up appending yet another instance.